rust/catalyst-contest/src/checkpoint/mod.rs

```diff
@@ -0,0 +1,454 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CatalystBallotCheckpointPayload {
+    /// What stage in the ballot processing is represented by this checkpoint.
+    stage: BallotProcessingStage,
+    /// Blake3 256‑bit digest of the root of the Sparse Merkle Tree (SMT) containing all
+    /// accepted ballot `document_ref`s up to and including this checkpoint.
+    smt_root: SmtRoot,
+    /// The total number of documents (leaves) in the SMT at this checkpoint.
+    smt_entries: SmtEntries,
+    /// Optional map of rejected contest ballots by reason.
+    rejections: Option<Rejections>,
+    /// Placeholder map of `document_ref => encrypted-tally-proposal-result`.
+    encrypted_tally: Option<EncryptedTally>,
+    /// Placeholder map of `document_ref => tally-proposal-result` for clear tally
+    /// snapshots.
+    tally: Option<Tally>,
+    /// Placeholder for D-Rep Encryption Key to allow decryption where required for audit
+    /// or published results.
+    drep_encryption_key: Option<DrepEncryptionKey>,
+}
+
+impl Encode<()> for CatalystBallotCheckpointPayload {
+    #[allow(clippy::arithmetic_side_effects)]
+    fn encode<W: minicbor::encode::Write>(
+        &self,
+        e: &mut minicbor::Encoder<W>,
+        ctx: &mut (),
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        let mut field_count = REQUIRED_FIELD_COUNT; // stage, smt-root, smt-entries are required
+        if self.rejections.is_some() {
+            field_count += 1;
+        }
+        if self.encrypted_tally.is_some() {
+            field_count += 1;
+        }
+        if self.tally.is_some() {
+            field_count += 1;
+        }
+        if self.drep_encryption_key.is_some() {
+            field_count += 1;
+        }
+
+        e.map(field_count)?;
+
+        e.str("stage")?;
+        self.stage.encode(e, ctx)?;
+
+        e.str("smt-root")?;
+        self.smt_root.encode(e, ctx)?;
+
+        e.str("smt-entries")?;
+        self.smt_entries.encode(e, ctx)?;
+
+        if let Some(ref rejections) = self.rejections {
+            e.str("rejections")?;
+            rejections.encode(e, ctx)?;
+        }
+
+        if let Some(ref encrypted_tally) = self.encrypted_tally {
+            e.str("encrypted-tally")?;
+            encrypted_tally.encode(e, ctx)?;
+        }
+
+        if let Some(ref tally) = self.tally {
+            e.str("tally")?;
+            tally.encode(e, ctx)?;
+        }
+
+        if let Some(ref drep_encryption_key) = self.drep_encryption_key {
+            e.str("drep-encryption-key")?;
+            drep_encryption_key.encode(e, ctx)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Decode<'_, ()> for CatalystBallotCheckpointPayload {
+    #[allow(clippy::arithmetic_side_effects)]
+    fn decode(
+        d: &mut minicbor::Decoder<'_>,
+        ctx: &mut (),
+    ) -> Result<Self, DecodeError> {
+        const MAX_FIELDS: u64 = REQUIRED_FIELD_COUNT + OPTIONAL_FIELD_COUNT;
+        let Some(map_len) = d.map()? else {
```

Comment:
No no, we should use `cbork-utils::map::Map` which would enforce all correct deterministic decoding checks.
You could take a look how it is used for `DocLocator` type as a reference.
```rust
let entries = Map::decode(d, &mut DecodeCtx::Deterministic)?;
```